nimbus/src/enrollment.rs

```diff
@@ -20,28 +24,367 @@ pub enum EnrolledReason {
+                    self.clone()
+                } else {
+                    let updated_enrollment = evaluate_enrollment(
+                        nimbus_id,
+                        available_randomization_units,
+                        app_context,
+                        updated_experiment,
+                    )?;
+                    log::debug!(
+                        "Experiment '{}' with enrollment {:?} is now {:?}",
+                        &self.slug,
+                        &self,
+                        updated_enrollment
+                    );
+                    if let EnrollmentStatus::Enrolled {
+                        branch,
+                        enrollment_id,
+                        ..
+                    } = &updated_enrollment.status
+                    {
+                        enrollment_events.push(EnrollmentChangeEvent::new(
+                            &updated_enrollment.slug,
+                            enrollment_id,
+                            branch,
+                            None,
+                            EnrollmentChangeEventType::Enrollment,
+                        ));
+                    };
+                    updated_enrollment
+                }
+            }
+            EnrollmentStatus::Enrolled {
+                ref branch,
+                enrollment_id,
+                ..
+            } => {
+                if !is_user_participating {
+                    log::debug!(
+                        "Existing experiment enrollment '{}' is now disqualified (global opt-out)",
+                        &self.slug
+                    );
+                    enrollment_events.push(EnrollmentChangeEvent::new(
+                        &self.slug,
+                        &enrollment_id,
+                        &branch,
+                        Some("optout"),
+                        EnrollmentChangeEventType::Disqualification,
+                    ));
+                    Self {
+                        slug: self.slug.clone(),
+                        status: EnrollmentStatus::Disqualified {
+                            reason: DisqualifiedReason::OptOut,
+                            enrollment_id,
+                            branch: branch.clone(),
+                        },
+                    }
+                } else {
+                    let updated_enrollment = evaluate_enrollment(
+                        nimbus_id,
+                        available_randomization_units,
+                        app_context,
+                        updated_experiment,
+                    )?;
+                    match updated_enrollment.status {
+                        EnrollmentStatus::Error { .. } => updated_enrollment,
+                        EnrollmentStatus::Enrolled {
+                            branch: ref new_branch,
+                            ..
+                        } if new_branch == branch => self.clone(),
+                        _ => {
```

Comment:
IIUC this will disqualify the user with `DisqualifiedReason::NotTargeted` under any of the following circumstances:

* The targeting expression no longer matches the user
    * (here `status` would match `EnrollmentStatus::NotEnrolled { reason: NotEnrolledReason::NotTargeted, .. }`)
* The bucketing params have changed and the user is no longer in one of the branches
    * (here `status` would match `EnrollmentStatus::NotEnrolled { reason: NotEnrolledReason::NotSelected, .. }`)
* The bucketing params have changed and the user is now assigned to a different branch than before
    * (here `status` would match `EnrollmentStatus::Enrolled { branch: new_branch, .. }` but `new_branch != branch`)
* There was an error while evaluating the updated experiment data
    * (here `status` would match `EnrollmentStatus::Error`)

I think that only the first of these should produce `DisqualifiedReason::NotTargeted`. My understanding is that bucketing changes should just be ignored and they stay in their existing `Enrolled` state and branch, emitting no events. For the error case, perhaps we should introduce a new `DisqualifiedReason::Error` variant to report this?

Reply:
This makes sense, I also wonder if we should do something if the branch we were in disappeared.

Reply:
Oh yes, good catch! I think `Disqualified` would also make sense in that case.